source/store.go

```diff
@@ -16,31 +16,118 @@ limitations under the License.
 
 package source
 
-import "fmt"
+import (
+	"errors"
+	"net/http"
+	"os"
+	"strings"
 
-var store = map[string]Source{}
+	"sync"
 
-// Register registers a Source under a given name.
-func Register(name string, source Source) {
-	store[name] = source
+	log "github.com/Sirupsen/logrus"
+	"github.com/linki/instrumented_http"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// ErrSourceNotFound is returned when a requested source doesn't exist.
+var ErrSourceNotFound = errors.New("source not found")
+
+// Config holds shared configuration options for all Sources.
+type Config struct {
+	Namespace     string
+	FQDNTemplate  string
+	Compatibility string
 }
 
-// Lookup returns a Source by the given name.
-func Lookup(name string) Source {
-	return store[name]
+// SingletonClientProvider provides singleton clients
+type SingletonClientProvider interface {
+	KubeClient() (kubernetes.Interface, error)
 }
 
-// LookupMultiple returns multiple Sources given multiple names.
-func LookupMultiple(names []string) ([]Source, error) {
-	sources := []Source{}
+// ClientProvider stores provider clients and guarantees that only one instance of client
+// will be generated
+type ClientProvider struct {
+	KubeConfig string
+	KubeMaster string
+	client     kubernetes.Interface
+	sync.Once
+}
+
+// KubeClient generates a kube client if it was not created before
+func (p *ClientProvider) KubeClient() (kubernetes.Interface, error) {
+	var err error
+	p.Once.Do(func() {
+		p.client, err = NewKubeClient(p.KubeConfig, p.KubeMaster)
+	})
+	return p.client, err
+}
 
+// ByNames returns multiple Sources given multiple names.
+func ByNames(p SingletonClientProvider, names []string, cfg *Config) ([]Source, error) {
```

Comment:
shouldn't we switch the name here? `ClientProvider` is the interface and `Singleton...`-something is the implementation.